### ISSUE TITLE
`endInTransaction` covers exceptions with exceptions from finally

### ISSUE THREAD

**Issue by petomat** [slick/slick#1822]
First of all I am using slick 3.1.0 and sqlite-jdbc-3.8.10.2 (probably more recent version don't differ that much, i.e. not in the related stack traces, this should be checked).
Just recently I ran into the following exception when finishing an transaction:

```java
java.sql.SQLException: cannot commit - no transaction is active
  at org.sqlite.core.DB.throwex(DB.java:859)
  at org.sqlite.core.DB.exec(DB.java:142)
  at org.sqlite.jdbc3.JDBC3Connection.setAutoCommit(JDBC3Connection.java:155)
  at slick.jdbc.JdbcBackend$BaseSession.endInTransaction(JdbcBackend.scala:443)
  at slick.driver.JdbcActionComponent$Commit$.run(JdbcActionComponent.scala:48)
  at slick.driver.JdbcActionComponent$Commit$.run(JdbcActionComponent.scala:46)
  at slick.dbio.SynchronousDatabaseAction$$anon$11.run(DBIOAction.scala:554)
  at slick.backend.DatabaseComponent$DatabaseDef$$anon$2.liftedTree1$1(DatabaseComponent.scala:237)
  at slick.backend.DatabaseComponent$DatabaseDef$$anon$2.run(DatabaseComponent.scala:237)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
  at java.lang.Thread.run(Thread.java:818)
```

Analyzing the stacktrace revealed that slick is executing a `slick.driver.JdbcActionComponent.Commit` like that

```scala
def run(ctx: Backend#Context): Unit = {
  try ctx.session.endInTransaction(ctx.session.conn.commit()) finally ctx.unpin
} 
```

and I think there is an exception happening while committing. The connection commit code looks like the following:

```java
public void commit() throws SQLException {
  checkOpen();
  if (autoCommit)
  throw new SQLException("database in auto-commit mode");
  db.exec("commit;");
  db.exec(beginCommandMap.get(transactionMode));
}
```

My strong assumption here ist that this exception is raised during `db.exec("commit;");` and thus before `db.exec(beginCommandMap.get(transactionMode));` so that in some way the transaction is committed but no new one is started via the command map.
Having said that, this exception is propagated from the above run method to the following call site:

```scala
private[slick] def endInTransaction(f: => Unit): Unit = {
  inTransactionally -= 1
  if(!isInTransaction) try f finally conn.setAutoCommit(true)
}
```

So that `f` throws that exception which is not caught, but the finally clause is executed, i.e. `setAutoCommit`:

```java
public void setAutoCommit(boolean ac) throws SQLException {
  checkOpen();
  if (autoCommit == ac)
  return;
  autoCommit = ac;
  db.exec(autoCommit ? "commit;" : beginCommandMap.get(transactionMode));
}
```

Here `autoCommit` is false since we are in a transaction but `ac` is true. Now `autoCommit` is set to true and another commit is executed on the db. But as described above the database is in a state were we have executed a commit but not started a new transaction, thus two commits are happening in a row now.
This results in a:

```java
java.sql.SQLException: cannot commit - no transaction is active
```

**TLDR:** I think the line `if(!isInTransaction) try f finally conn.setAutoCommit(true)` covers exceptions that happen in `f` with exceptions happening in the finally clause. This may be an undesired behavior since the call site might check for the former exception type.

Some relevant links: 

https://www.javamex.com/tutorials/exceptions/exceptions_finally.shtml
https://softwareengineering.stackexchange.com/questions/188858/throwing-an-exception-inside-finally

Another side note: I found it somewhat disturbing that the auto-commit flag is reset to true regardless of the value before the transaction was started.

**Comment by hvesalai** [comment]
If you can, please provide a pull request to fix this.
